penrose-web/src/compiler/Style.ts

```diff
@@ -1,7 +1,2068 @@
+const insertLabels = (trans: Translation, labels: LabelMap): Translation => {
+  console.log("labels", labels);
+
+  const insertLabel = (name: string, fieldDict: FieldDict): [string, FieldDict] => {
+    let label: string = labels[name];
+    if (!label) {
+      label = name;
+      // Note that there's no reason for a shape to have no label; by default, it will be its substance name
+    }
+
+    // COMBAK: Model this better WRT Maybes; need to distinguish between "no label mapping" and "no label" so we can delete the relevant text GPIs (see Haskell for code)
+    fieldDict[LABEL_FIELD] = {
+      tag: "FExpr",
+      contents: {
+        tag: "Done",
+        contents: {
+          tag: "StrV",
+          contents: label
+        }
+      }
+    };
+    return [name, fieldDict];
+  };
+
+  return mapTrans(trans, insertLabel);
+};
+
+const translateStyProg = (varEnv: Env, subEnv: SubEnv, subProg: SubProg, styProg: StyProg, labelMap: LabelMap, styVals: number[]): Either<StyErrors, Translation> => {
+  // COMBAK: Deal with styVals
+
+  const res = foldM(styProg.blocks, (trans, hb, i) => translatePair(varEnv, subEnv, subProg, trans, hb, i), initTrans());
+
+  if (isLeft(res)) { return res; } // Return errors
+
+  const trans = res.contents;
+  const transWithNames = insertNames(trans);
+  const transWithNamesAndLabels = insertLabels(transWithNames, labelMap); // TODO
+
+  // COMBAK: Do this with plugins
+  // const styValMap = styJsonToMap(styVals);
+  // const transWithPlugins = evalPluginAccess(styValMap, transWithNamesAndLabels);
+  // return Right(transWithPlugins);
+  return Right(transWithNamesAndLabels);
+};
+
+//#endregion
+
+// BEGIN GENOPTPROBLEM.HS PORT
+
+//#region Translation utilities -- TODO move to EngineUtils
+
+function foldFields<T>(f: (s: string, field: Field, fexpr: FieldExpr<VarAD>, acc: T[]) => T[],
+  [name, fieldDict]: [string, { [k: string]: FieldExpr<VarAD> }], acc: T[]): T[] {
+
+  const res: T[] = Object.entries(fieldDict).reduce((acc: T[], [field, expr]) => f(name, field, expr, acc), []);
+  return res.concat(acc);
+};
+
+function foldSubObjs<T>(f: (s: string, f: Field, fexpr: FieldExpr<VarAD>, acc: T[]) => T[], tr: Translation): T[] {
+  return Object.entries(tr.trMap).reduce((acc: T[], curr) => foldFields(f, curr, acc), []);
+};
+
+//#endregion
+
+//#region Gen opt problem
+
+// Find varying (float) paths
+// For now, don't optimize these float-valued properties of a GPI
+// (use whatever they are initialized to in Shapes or set to in Style)
+const unoptimizedFloatProperties: String[] =
+  ["rotation", "strokeWidth", "thickness", "transform", "transformation",
+    "opacity", "finalW", "finalH", "arrowheadSize"]
+
+const optimizedVectorProperties: string[] =
+  ["start", "end", "center"]
+
+const declaredVarying = (t: TagExpr<VarAD>): boolean => {
+  if (t.tag === "OptEval") {
+    if (t.contents.tag === "Vary") {
+      return true;
+    }
+  }
+
+  return false;
+};
+
+const mkPath = (strs: string[]): Path => {
```

Comment:
Instead of constructing dummy nodes, consider creating a hash or just use a concatenated string (if they are guaranteed to be unique) of the paths. The AST node types are really meant to be _just for the parser/checker phase_. These big `ASTNode`s are probably not great for lookups and are prone to errors.